### PR TITLE
ekf2: EV overhaul yaw and position fusion

### PR DESCRIPTION
**Pull request by dagar**
- move EV yaw and EV position to new state machines
 - new EV position offset estimator
 - yaw_align now strictly means north (no more rotate external vision aid mask)
 - automatically switching between EV yaw, and yaw align north based on GPS quality
 - ~~requires https://github.com/PX4/PX4-Autopilot/pull/20498~~
 - broken out of https://github.com/PX4/PX4-Autopilot/pull/19128

### CODE REVIEW

File: src/modules/ekf2/EKF/gps_control.cpp [hunk 1]
```diff
@@ -81,12 +83,34 @@ void Ekf::controlGpsFusion()
...
 						     pos_obs_var,                                // observation variance
 						     math::max(_params.gps_pos_innov_gate, 1.f), // innovation gate
 						     _aid_src_gnss_pos);
 		_aid_src_gnss_pos.fusion_enabled = (_params.gnss_ctrl & GnssCtrl::HPOS);
 
+		// if GPS is otherwise ready to go, but yaw_align is blocked by EV give mag a chance to start
+		if (_control_status.flags.tilt_align && _NED_origin_initialised
+		    && gps_checks_passing && !gps_checks_failing) {
+
+			if (!_control_status.flags.yaw_align) {
```
Comment (dagar) [inline src/modules/ekf2/EKF/gps_control.cpp:96]: @bresch FYI in testing I also had the yaw estimator able to yaw_align here (eg transitioning from indoor -> outdoor), but I wanted to untangle more of mag control (https://github.com/PX4/PX4-Autopilot/pull/20470) before pursuing that upstream.

File: src/modules/ekf2/EKF/mag_control.cpp [hunk 2]
```diff
@@ -303,7 +311,7 @@ void Ekf::checkYawAngleObservability()
 				: _accel_lpf_NE.norm() > 2.0f * _params.mag_acc_gate;
 
 	_yaw_angle_observable = _yaw_angle_observable
-				&& (_control_status.flags.gps || _control_status.flags.ev_pos); // Do we have to add ev_vel here?
+				&& (_control_status.flags.gps || (_control_status.flags.ev_pos && _control_status.flags.yaw_align));
```
Comment (bresch) [inline src/modules/ekf2/EKF/mag_control.cpp:314]: The yaw angle is only observable if the ev_pos data is in the local frame, if the data is in body frame and that the yaw is aligned using the mag, the yaw isn't observable.

**Review by bresch — CHANGES_REQUESTED** [review]
Mostly good but I find some parts of the code too hard to follow with those deeply nested `if` statements containing a lot of code

File: src/modules/ekf2/EKF/ev_pos_control.cpp [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+						_information_events.flags.reset_pos_to_vision = true;
+
+						if (_control_status.flags.gps) {
+							resetHorizontalPositionTo(measurement - _ev_pos_b_est.getBias(), measurement_var + _ev_pos_b_est.getBiasVar());
+							_ev_pos_b_est.setBias(-Vector2f(_state.pos.xy()) + measurement);
+
+						} else {
+							resetHorizontalPositionTo(measurement, measurement_var);
+							_ev_pos_b_est.reset();
+						}
```
[the remark this answers is not in the document]
Comment (dagar) [reply src/modules/ekf2/EKF/ev_pos_control.cpp:237]: It's actually more about context and how slightly different every single `resetHorizontalPositionToVision` actually is. The reset we do here (due to fusion failing) is slightly different than the reset we do when starting EV pos or when EV itself indicates it has reset.

I do agree we need to do more to cleanly split up the code sprawl, I would just rather keep it as dumb flat code until we can figure out how to do that without lossy abstractions.

File: src/modules/ekf2/EKF/ev_pos_control.cpp [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+
+		if (continuing_conditions_passing) {
+
+			const bool bias_estimator_change = (bias_fusion_was_active != _ev_pos_b_est.fusionActive());
+
+			if (ev_reset || yaw_alignment_changed || bias_estimator_change) {
+
+				if (quality_sufficient) {
+
+					if (!_control_status.flags.gps) {
```
[the remark this answers is not in the document]
Comment (dagar) [reply src/modules/ekf2/EKF/ev_pos_control.cpp:188]: Yes...

File: src/modules/ekf2/EKF/mag_control.cpp (hunk 2 above)
Comment (dagar) [reply src/modules/ekf2/EKF/mag_control.cpp:314]: Yes that's the intent here, and you're obviously correct about the potential problem, but I believe it's not really an issue due to how intertwined the control logic is between EV yaw/pos <=> mag <=> GPS. This is related to the scattered/distributed state machine issues I mentioned a few months ago.

We *should* only have ev_pos + yaw_align if the incoming EV is NED (or at least tells us it is...), however I agree from these flags along it's a little ambiguous.

Assuming EV data is an arbitrary world heading (FRD)
 - if GPS is bad/inactive
   - EV yaw is allowed to start if GPS isn't active (_control_status.flags.yaw_align = false)
   - EV yaw will stop all mag fusion
   - EV position fusion using data directly (no bias estimate, no rotation, etc) 
 - GPS good/active
   -  EV yaw continuing conditions no longer passing
   - mag fusion no longer blocked
   - EV position rotated into EKF frame and position bias estimator activated

Let me think about a proper way to capture and enforce this in the code.

**Review by bresch — APPROVED** [review]

File: src/modules/ekf2/test/test_EKF_externalVision.cpp [hunk 5]
```diff
@@ -254,9 +253,9 @@ TEST_F(EkfExternalVisionTest, visionAlignment)
 	EXPECT_TRUE(velVar_new(1) > velVar_new(0));
 
 	// THEN: the frame offset should be estimated correctly
-	Quatf estimatedExternalVisionFrameOffset = _ekf->getVisionAlignmentQuaternion();
-	EXPECT_TRUE(matrix::isEqual(externalVisionFrameOffset.canonical(),
-				    estimatedExternalVisionFrameOffset.canonical()));
+	// Quatf estimatedExternalVisionFrameOffset = _ekf->getVisionAlignmentQuaternion();
```
Comment (bresch) [inline src/modules/ekf2/test/test_EKF_externalVision.cpp:256]: Add a note to explain why this doesn't work or remove it?

**Comment by hamishwillee** [comment]
This removed `MAV_ODOM_LP` which is referenced in the [VIO guide](https://docs.px4.io/main/en/computer_vision/visual_inertial_odometry.html#check-verify-vio-estimate) as part of the instructions for verifying the estimate:

> Set the PX4 parameter MAV_ODOM_LP to `1``. PX4 will then stream back the received external pose as MAVLink [ODOMETRY (opens new window)](https://mavlink.io/en/messages/common.html#ODOMETRY)messages. You can check these MAVLink messages with the QGroundControl [MAVLink Inspector](https://docs.qgroundcontrol.com/master/en/qgc-user-guide/analyze_view/mavlink_inspector.html)

Is there anyone who can help fix up the documentation/explain how we check this now? Is the best fix just to delete the section? @dagar @beniaminopozzan Can you advise?

This comes from https://github.com/PX4/PX4-user_guide/issues/3094

**Comment by beniaminopozzan** [comment]
Hi @hamishwillee , apologies but I can't work on this right now.
As far as I remember, that guide is still valid up to the checking procedure. I would temporarily mark that last section as "working up to `v1.13.x`.

**Comment by hamishwillee** [comment]
Thanks @beniaminopozzan - much appreciated. I added a note in https://github.com/PX4/PX4-user_guide/pull/3138

Is there someone else you can recommend look at this, or some way you can track it to look at it later? What I'm asking is what is the best way to record this action - just create an issue and hope someone takes it?
